src/nti/transactions/loop.py

```diff
@@ -127,11 +127,23 @@ def __init__(self, response, reason):
     #: Deprecated alias for `AbortAndReturn`
     AbortException = AbortAndReturn
 
-    #: If not none, this should be a number that will be passed to
-    #: time.sleep or gevent.sleep in between retries.
-    #: TODO: add a backoff delay interval/multiplier between retries.
+    #: If not None, this is a number that specifies the base amount of time
+    #: to wait after a failed transaction attempt before retrying. Each
+    #: retry attempt will pick a delay following the binary exponential
+    #: backoff algorithm: ``sleep * (random number between 0 and 2^retry-1)``.
+    #: (A simple increasing multiplier might work well if there is only one other
+    #: transaction that we conflict with, but in cases of multiple conflicts
+    #: or even new conflicts, the random backoff should provide higher
+    #: overall progress.)
+    #: This can be set in a subclass if not passed to the constructor.
     sleep = None
+
+    #: How many total attempts will be made, including the initial. This
+    #: can be set in a subclass, if not passed to the constructor.
     attempts = 10
```

Comment:
Are ``sleep = None`` and ``attempts = 10`` still the most appropriate defaults?

Reply:
I suspect that attempts is too high for most uses. `transaction` and `pyramid_retry` both use 3. I was planing to change to that in a subsequent PR.

As for sleep, I think that's a bit more complicated.  In this generic loop, I think not sleeping is probably appropriate because we don't have any expectations around concurrency. But in the Pyramid subclass, where we do have expectations around highly concurrent usage, I think a non-zero default value is probably appropriate. If we go with three retries, then in the worst case that we actually have to take those retries, our expected total backoff time would be `0.5 * sleep + 1.5 * sleep + 3.5 * sleep = 5.5 * sleep` . We want to choose a value that minimizes conflicts while also minimizing delays; that's hard to do statically. 

I mentioned this on another ticket and dismissed it as unnecessary complexity, but rethinking it, what if the base sleep value was dynamic for each request? Specifically, what if we derive `sleep` from the amount of time the request ran? `sleep = min(min(X% of run time, Y), Z)`, where X, Y, and Z would be configurable, defaulting to, say 10%, 10ms and 50ms. The idea is that the more expensive the request was to run, the longer we should be willing to wait to retry it (because the longer we wait, the more likely it is that other concurrent processes we're conflicting with will have finished). But we still don't want to wait too long because (a) the user is waiting and (b) the longer we wait the more likely it is that we may have to actually do more work to achieve the desired result.

Reply:
Moved to https://github.com/NextThought/nti.transactions/issues/34